crates/astria-composer/src/searcher/executor.rs

```diff
@@ -120,178 +139,266 @@ impl Executor {
-        let rsp = self
-            .sequencer_client
-            .submit_transaction_sync(signed_tx)
-            .await
-            .wrap_err("failed submitting transaction to sequencer")?;
-        if rsp.code.is_err() {
-            bail!(
-                "submitting transaction to sequencer returned with error code; code: `{code}`; \
-                 log: `{log}`; hash: `{hash}`",
-                code = rsp.code.value(),
-                log = rsp.log,
-                hash = rsp.hash,
-            );
-        }
-        Ok(())
-    }
-
-    /// Signs and submits the bundle of actions to the sequencer.
-    async fn sign_and_submit(&mut self, actions: Vec<Action>) -> eyre::Result<()> {
-        let nonce = self
-            .get_and_increment_nonce()
-            .ok_or(eyre!("no nonce stored; cannot process bundle"))?;
-
-        let tx = UnsignedTransaction {
-            nonce,
-            actions,
-        }
-        .into_signed(&self.sequencer_key);
-
-        self.submit_tx(tx)
-            .await
-            .wrap_err("failed submitting signed actions to sequencer")?;
-        Ok(())
-    }
-
     /// Run the Executor loop, calling `process_bundle` on each bundle received from the channel.
     ///
     /// # Errors
     /// An error is returned if connecting to the sequencer fails.
     pub(super) async fn run_until_stopped(mut self) -> eyre::Result<()> {
-        // set up connection to sequencer
-        self.init_nonce_from_sequencer(5, Duration::from_secs(5), 2.0)
+        let mut submission_fut = Fuse::terminated();
+        self.nonce = get_latest_nonce(self.sequencer_client.clone(), self.address)
             .await
-            .wrap_err("failed retrieving initial nonce from sequencer")?;
+            .wrap_err("failed getting initial nonce from sequencer")?;
+        self.status.send_modify(|status| status.is_connected = true);
+        loop {
+            select! {
+                biased;
 
-        while let Some(bundle) = self.executor_rx.recv().await {
-            if let Err(e) = self.sign_and_submit(bundle).await {
-                // FIXME: currently this will fail both when there is an issue with the nonce and
-                // when unable to reach the sequencer. As there is currently no error returned by
-                // the sequencer for invalid nonces, there is nothing to handle. This should be
-                // changed after #364 is merged in a followup PR to handle nonce failues.
-                error!(
-                    error.message = %e,
-                    error.cause_chain = ?e,
-                    "failed submitting the bundle to the sequencer; bailing",
-                );
-                break;
+                rsp = &mut submission_fut, if !submission_fut.is_terminated() => {
+                    match rsp {
+                        Ok(new_nonce) => self.nonce = new_nonce,
+                        Err(_e) => bail!("eyeyey"),
```

Comment:
???

Reply:
:( ty for flagging this. I overlooked it

Reply:
I am not properly returning the error. I now also distinguish between "submission" as the whole process of attaching nonce + signing + sending, and just "sending" as the last step of that process. That way it's clearer what failed.